typescript/src/proof.ts

```diff
@@ -70,3 +79,253 @@ function createMerkleProof(txMerkleBranch: TransactionMerkleBranch): string {
+ * @param previousEpochDifficulty The difficulty of the previous Bitcoin epoch.
+ * @param currentEpochDifficulty The difficulty of the current Bitcoin epoch.
+ * @dev The block headers must come from Bitcoin epochs with difficulties marked
+ *      by the previous and current difficulties. If a Bitcoin difficulty relay
+ *      is used to provide these values and the relay is up-to-date, only the
+ *      recent block headers will pass validation. Block headers older than the
+ *      current and previous Bitcoin epochs will fail.
+ * @throws {Error} If any of the block headers are invalid, or if the block
+ *         header chain is not continuous.
+ * @returns An empty return value.
+ */
+function validateBlockHeadersChain(
+  blockHeaders: BlockHeader[],
+  previousEpochDifficulty: BigNumber,
+  currentEpochDifficulty: BigNumber
+) {
+  let requireCurrentDifficulty: boolean = false
+  let previousBlockHeaderHash: Hex = Hex.from("00")
+
+  for (let index = 0; index < blockHeaders.length; index++) {
+    const currentHeader = blockHeaders[index]
+
+    // Check if the current block header stores the hash of the previously
+    // processed block header. Skip the check for the first header.
+    if (index !== 0) {
+      if (
+        !previousBlockHeaderHash.equals(currentHeader.previousBlockHeaderHash)
+      ) {
+        throw new Error("Invalid headers chain")
+      }
+    }
+
+    const difficultyTarget = bitsToDifficultyTarget(currentHeader.bits)
+
+    const currentBlockHeaderHash = computeHash256(
+      serializeBlockHeader(currentHeader)
+    )
+
+    // Ensure the header has sufficient work.
+    if (
+      hashLEToBigNumber(currentBlockHeaderHash.toString()).gt(difficultyTarget)
+    ) {
+      throw new Error("Insufficient work in the header")
+    }
+
+    // Save the current block header hash to compare it with the next block
+    // header's previous block header hash.
+    previousBlockHeaderHash = currentBlockHeaderHash
+
+    // Check if the stored block difficulty is equal to previous or current
+    // difficulties.
+    const difficulty = targetToDifficulty(difficultyTarget)
+
+    if (previousEpochDifficulty.eq(1) && currentEpochDifficulty.eq(1)) {
+      // Special case for Bitcoin Testnet. Do not check block's difficulty
+      // due to required difficulty falling to `1` for Testnet.
+      continue
+    }
+
+    if (
+      !difficulty.eq(previousEpochDifficulty) &&
+      !difficulty.eq(currentEpochDifficulty)
+    ) {
+      throw new Error(
+        "Header difficulty not at current or previous Bitcoin difficulty"
+      )
+    }
+
+    // Additionally, require the header to be at current difficulty if some
+    // headers at current difficulty have already been seen. This ensures
+    // there is at most one switch from previous to current difficulties.
+    if (requireCurrentDifficulty && !difficulty.eq(currentEpochDifficulty)) {
+      throw new Error("Header must be at current Bitcoin difficulty")
+    }
```

Comment:
I see we are doing it differently than the SPV solidity lib does (https://github.com/keep-network/tbtc-v2/blob/8b2b2b97c856b4bed42a97788cfa2af8b6aba8d8/solidity/contracts/bridge/BitcoinTx.sol#L227). We are checking headers separately while the lib checks the accumulated difficulty. Could you explain the exact reasons?

Reply:
In the solidity implementation there can be a situation when the current difficulty is lower than the previous difficulty and the first header is in the previous difficulty epoch and subsequent headers are in the current difficulty epoch.
In that case, the total accumulated difficulty will be lower than the required difficulty and the validation will fail.
Therefore, we had to either ensure all the headers in the chain are in the same epoch or we must provide a longer headers chain.

This typescript implementation is stricter and more precise. We simply verify the difficulty of each header separately: the difficulty must be at previous or current difficulty. Additionally, while iterating over the headers we ensure that the possible switch from previous difficulty headers and current difficulty ones happens at most one time. We can also precisely verify that the provided headers chain's length equals exactly the required number of confirmations.

Reply:
Makes sense! Thanks!